codalab/model/worker_model.py

```diff
@@ -0,0 +1,322 @@
+                                cl_worker_socket.c.worker_id == worker_id))
+            ).fetchall()
+            for socket_row in socket_rows:
+                self._cleanup_socket(socket_row.socket_id)
+            conn.execute(cl_worker_socket.delete()
+                         .where(and_(cl_worker_socket.c.user_id == user_id,
+                                     cl_worker_socket.c.worker_id == worker_id)))   
+            conn.execute(cl_worker_run.delete()
+                         .where(and_(cl_worker_run.c.user_id == user_id,
+                                     cl_worker_run.c.worker_id == worker_id)))      
+            conn.execute(cl_worker_dependency.delete()
+                         .where(and_(cl_worker_dependency.c.user_id == user_id,
+                                     cl_worker_dependency.c.worker_id == worker_id)))      
+            conn.execute(cl_worker.delete()
+                         .where(and_(cl_worker.c.user_id == user_id,
+                                     cl_worker.c.worker_id == worker_id)))
+            
+    def get_workers(self):
+        """
+        Returns information about all the workers in the database.
+        """
+        with self._engine.begin() as conn:
+            worker_rows = conn.execute(cl_worker.select()).fetchall()
+            worker_run_rows = (
+                conn.execute(cl_worker_run.select()).fetchall())
+            worker_dependency_rows = (
+                conn.execute(cl_worker_dependency.select()).fetchall())
+ 
+        worker_dict = {(row.user_id, row.worker_id): {
+            'user_id': row.user_id,
+            'worker_id': row.worker_id,
+            'slots': row.slots,
+            'checkin_time': row.checkin_time,
+            'socket_id': row.socket_id,
+            'run_uuids': [],
+            'dependency_uuids': [],
+        } for row in worker_rows}
+        for row in worker_run_rows:
+            worker_dict[(row.user_id, row.worker_id)]['run_uuids'].append(row.run_uuid)
+        for row in worker_dependency_rows:
+            worker_dict[(row.user_id, row.worker_id)]['dependency_uuids'].append(row.dependency_uuid)
+        return worker_dict.values()
+
+    def get_bundle_worker(self, uuid):
+        """
+        Returns information about the worker that the given bundle is running
+        on. This method should be called only for bundles that are running.
+        """
+        with self.engine.begin() as connection:
+            row = connection.execute(cl_worker_run.select()
+                                     .where(cl_worker_run.c.run_uuid == uuid)).fetchone()
+            precondition(row, 'Trying to find worker for bundle that is not running.')
+            if row:
+                return {
+                    'user_id': row.user_id,
+                    'worker_id': row.worker_id,
+                    'socket_id': row.socket_id,
+                }
+
+    def allocate_socket(self, user_id, worker_id, conn=None):
```

Comment:
This method is public (doesn't have the _ prefix). It's called in code that makes read requests to the worker. Those don't need to do other database requests in the same transaction.